### PR TITLE
Enforce limits on tile GPU memory usage

### PR DESCRIPTION
**Pull request by pmconne**
Some users are trying to view massive iModels on laptops with poor integrated graphics chips and experiencing WebGL context loss due to exhaustion of GPU memory. Most of the memory is being consumed by tiles that are not currently being drawn. In particular, we tend to keep a parent tile's graphics in memory as long as its children are in memory - so if you zoom in far enough, lots of parent tiles are hogging memory despite not being drawn.

`TileAdmin.maxTotalContentBytes`, if defined, specifies the maximum amount of GPU memory that should be allocated to tile contents. When this  threshold is exceeded, we dispose of tile contents in least-recently-selected order until either the limit is satisfied or we have discarded the contents of all non-selected tiles. (Here "selected" means "selected for display in any viewport").

To make this work I needed to be able to know how much memory each Tile is consuming; how much total memory all Tiles are consuming; which tiles are currently selected for display in any viewport; and a least-to-most-recently-selected ordering of all tiles. I wanted to introduce as little overhead (in additional memory and in processing) as possible. The solution is an intrusive linked list - see documentation comments on `LRUTileList` for details.

Tile trees need to be aware that any Tile's content can become unloaded at any time due to memory pressure, and avoid implementing tile selection in such a way that requires parents to be loaded before their children can be displayed - that would defeat the purpose and end up constantly reloading the same tile content only for it to be disposed again. IModelTile required changes in this regard.

ImageryMapTiles are weird in that they use reference-counting; `disposeContents` does nothing if the ref count is > 0, instead relying on the garbage collector to free their memory. That's unfortunate but `LRUTileList` handles it by keeping such tiles in the list.

Incidental changes:
- TileAdmin was an abstract base class with one non-exported implementation. That was becoming unwieldy as every new addition to the base class interface had to be duplicated in the implementation. I collapsed them into one class.
- Added `OrderedSet<T>` which implements the interface of `Set<T>` with the custom ordering of `SortedArray<T>`.

### CODE REVIEW

File: core/frontend/src/tile/TileAdmin.ts [hunk 1]
```diff
@@ -54,189 +54,475 @@ export interface SelectedAndReadyTiles {
...
+  private readonly _selectedAndReady = new Map<Viewport, SelectedAndReadyTiles>();
+  private readonly _viewportSetsForRequests = new UniqueViewportSets();
+  private _maxActiveRequests: number;
+  private readonly _maxActiveTileTreePropsRequests: number;
+  private _defaultTileSizeModifier: number;
+  private readonly _retryInterval: number;
+  private readonly _enableInstancing: boolean;
+  private readonly _enableImprovedElision: boolean;
+  private readonly _ignoreAreaPatterns: boolean;
+  private readonly _enableExternalTextures: boolean;
```
Comment (pmconne) [inline core/frontend/src/tile/TileAdmin.ts:72]: @markschlosseratbentley please double-check I didn't mess something up when merging your external textures changes into TileAdmin. (Your integration still passes so probably ok).

Comment (markschlosseratbentley) [reply core/frontend/src/tile/TileAdmin.ts:72]: @pmconne I went over your code, and it looks good with relation to external textures.  I also built + ran.  External textures all seem in order & working.

**Comment by pmconne** [comment]
@RBBentley I had to disable this feature for MapTile and ImageryMapTile - with their dependence on one another and the reference-counting, there isn't really a way to determine if it's safe to discard their contents.

**Review by bbastings — COMMENTED** [review]
Hypermodeling changes approved.

**Comment by pmconne** [comment]
@markschlosseratbentley this supplants your `mobileExpirationThreshold` which I have removed - please review.

**Review by markschlosseratbentley — APPROVED** [review]
@pmconne 

This all looks good to me.  Nice stuff - I like the various profiles for thresholding the memory.

Importantly, I tried out your changes on iOS.  The problematic file that my original mobile threshold code fixed (CCNav reality model) suffers no context loss.  I pulled up the memory usage dialog in display-test-app and watched the memory never go above 200MB as I aggressively panned, rotated, and loaded new tiles.  That was the situation that lost context previously.

So this PR appears to still fix the iOS reality model context loss issue.

**Comment by markschlosseratbentley** [comment]
> @markschlosseratbentley this supplants your `mobileExpirationThreshold` which I have removed - please review.

@pmconne  See my approval comments.  (In summary: looks great and appears to still fix iOS issues that `mobileExpirationThreshold` resolved).

**Review by RBBentley — APPROVED** [review]
Wow -- excellent implementation.   I believe we could work map tiles into this scheme if necessary but hopfully their tile count is limited enough such that it is not necessary.

**Comment by pmconne** [comment]
@Mergifyio backport release/2.12.x

**Pull request merged**